### DISCUSSION TITLE
How to render styles for tests?

### DISCUSSION THREAD

**Discussion by lukemorales** [stitchesjs/stitches#800] [Help]
When using something like `styled-components` we're able to render a component for testing and then test if that component has some style based on the use-case we want to test. With stitches, the rendered component only has classNames attached, but no reference to what styles those classNames apply, so there's no way to test styles based on conditions.

Is there something we can do on a test renderer or on jest setup file to make those styles be applied to the components?

**Comment by peduarte** [comment]
Hey! What are you trying to test though? You shouldn't need to test that the styles are applied, that's Stitches' job. 

This repo does some basic tests: https://github.com/pffigueiredo/test-ui/blob/master/src/Button.test.tsx

**Reply by lukemorales** [reply]
@peduarte I'd like to assert if a certain style is present based on some prop, I might do some verification inside the component that might trigger a variant or another. Testing styles would also be valuable for cases where we aren't aware that we need to do a compound variant because one variant might be interfering with another variant styles.
Also, for design systems, I would say is reasonable that I'd also like to assert the styles for the default props. Not to test that Stitches is working, but as a way to make sure no one makes that change without consent.

Here's an example of some tests done with styled-components that would also be nice to be able to test with Stitches: https://github.com/Won-Games/client/blob/c1fc7174695fa172a2dfea02a9b5575c3a16ae1e/src/components/Button/test.tsx#L41-L44

**Reply by peduarte** [reply]
That makes sense. We'll need to think about how to do this. I will cc @jonathantneal

**Reply by peduarte** [reply]
BTW, if we have a clear idea of how to proceed we can create a RFC Discussion.

**Reply by jonathantneal** [reply]
> I'd like to assert if a certain style is present based on some prop

That’s very reasonable. Thanks for the example.

Does **jest** ship `toHaveStyle` out of the box? Do they support [`getComputedStyle`](https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle)? If so, we can document this pattern, and it isn’t anything we need to change about Stitches.

**Reply by lukemorales** [reply]
@jonathantneal `toHaveStyle` comes from [@testing-library/jest-dom](https://github.com/testing-library/jest-dom) and jest supports `getComputedStyle` via `jsdom`, and here are my tests with `getComputedStyle` but the output of it doesn't seem to be what we would expect:

<img width="1056" alt="CleanShot 2021-09-21 at 16 22 33@2x" src="https://user-images.githubusercontent.com/14251143/134234446-aed32d1f-56c1-472b-972b-669d5437c075.png">

This is what the console logs for the computed styles:

```js
    CSSStyleDeclaration {
      '0': 'font',
      '1': 'text-rendering',
      '2': 'letter-spacing',
      '3': 'word-spacing',
      '4': 'line-height',
      '5': 'text-transform',
      '6': 'text-indent',
      '7': 'text-shadow',
      '8': 'display',
      '9': 'text-align',
      '10': 'align-items',
      '11': 'cursor',
      '12': 'color',
      '13': 'padding',
      '14': 'border',
      '15': 'background-color',
      '16': 'box-sizing',
      '17': 'visibility',
      _values: {
        'font-family': '-webkit-small-control',
        font: '-webkit-small-control',
        'text-rendering': 'auto',
        'letter-spacing': 'normal',
        'word-spacing': 'normal',
        'line-height': 'normal',
        'text-transform': 'none',
        'text-indent': '0',
        'text-shadow': 'none',
        display: 'inline-block',
        'text-align': 'center',
        'align-items': 'flex-start',
        cursor: 'default',
        color: 'ButtonText',
        padding: '2px 6px 3px 6px',
        'padding-top': '2px',
        'padding-right': '6px',
        'padding-bottom': '3px',
        'padding-left': '6px',
        'border-top-width': '2px',
        'border-right-width': '2px',
        'border-bottom-width': '2px',
        'border-left-width': '2px',
        'border-width': '2px',
        'border-top-style': 'outset',
        'border-right-style': 'outset',
        'border-bottom-style': 'outset',
        'border-left-style': 'outset',
        'border-style': 'outset',
        'border-top-color': 'buttonface',
        'border-right-color': 'buttonface',
        'border-bottom-color': 'buttonface',
        'border-left-color': 'buttonface',
        'border-color': 'buttonface',
        border: '2px outset buttonface',
        'border-top': '2px outset buttonface',
        'border-left': '2px outset buttonface',
        'border-right': '2px outset buttonface',
        'border-bottom': '2px outset buttonface',
        'background-color': 'ButtonFace',
        'box-sizing': 'border-box',
        visibility: 'visible'
      },
      _importants: {
        margin: '',
        font: '',
        'text-rendering': '',
        color: '',
        'letter-spacing': '',
        'word-spacing': '',
        'line-height': '',
        'text-transform': '',
        'text-indent': '',
        'text-shadow': '',
        display: '',
        'text-align': '',
        'align-items': '',
        cursor: '',
        padding: '',
        border: '',
        'background-color': '',
        'box-sizing': '',
        visibility: undefined
      },
      _length: 18,
      _onChange: [Function (anonymous)]
    }
```

Maybe you're thinking of a different approach?

**Reply by jkytoela** [reply]
I'm having the same issues.

![kuva](https://user-images.githubusercontent.com/26070640/134296534-611b60c4-f524-46be-9037-d418a3e413e3.png)

**Reply by lukemorales** [reply]
@jonathantneal did you see our replies here? Just getting your attention again on this subject as there's no reply or reactions from you on the answers 😅

**Reply by jonathantneal** [reply]
Hi, @lukemorales. One of us should be able to investigate this relatively soon. Now that I know `jsdom` is involved, I have to suspect the issue is a lack of compatibility on their end (_where I use the term "compatibility" similar to describing compatibility with Internet Explorer_). The last time we investigated a `jest` issue involving `jsdom`, I wound up committing fixes to the `cssom` package that powered the `jsdom` package that powered `jest` (_and the `cssom` package is no longer actively maintained_).

**Reply by mryechkin** [reply]
Just chiming in on this - this exact use case is needed for us as well (testing defaults in a design system library). Currently we're using the same approach as @lukemorales above with `styled-components`, and would need to have similar tests after converting to Stitches.

**Reply by rogerluiz** [reply]
> Hey! What are you trying to test though? You shouldn't need to test that the styles are applied, that's Stitches' job.
> 
> This repo does some basic tests: https://github.com/pffigueiredo/test-ui/blob/master/src/Button.test.tsx

this is like a bot reply aways the same, if i want to test the style, that every library like styled-component do, the stitches library have to, what the purpose of not do that?

**Comment by gmlnchv** [comment]
Hi, @jonathantneal. Wondering if you or anyone else has had a chance to look into this. Having an issue where `getCssText()` is always an empty string. I've also tried plugging in a test from [here](https://github.com/modulz/stitches/blob/canary/packages/core/tests/theme.js#L4-L18) and everything is ok except `getCssText()` is an empty string:

<img width="800" alt="Screen Shot 2021-10-11 at 2 30 56 PM" src="https://user-images.githubusercontent.com/1403779/136721202-e5804807-924a-4f26-8913-556ed51e5172.png">

Any pointers appreciated!

**Comment by gmlnchv** [comment]
FWIW, testing CSS using `@playwright/test` and the `.toHaveCSS()` matcher that comes with it, solves the issue for me (obv being an actual browser environment, not `jsdom`).

```
test('has default styles', async ({ page }) => {
    await page.goto('http://localhost:6006/iframe.html?id=styles-box--default');
    const box = page.locator('.c-lesPJm');
    await expect(box).toHaveCSS('width', '64px');
    await expect(box).toHaveCSS('height', '64px');
    await expect(box).toHaveCSS('border-radius', '0px');
    await expect(box).toHaveCSS('background-color', 'rgb(18, 66, 109)');
  });
```

**Comment by mryechkin** [comment]
Hi folks. Has anyone had a chance to look into this further?

We'd like to start moving our component library to use Stitches (currently on `styled-components`), and being able to do the kinds of tests @lukemorales mentioned is crucial for us. In my preliminary testing I'm running into all of the above-mentioned issues.

Has anyone been able to find a suitable workaround with Jest?

**Comment by deanlot** [comment]
Hi all,

Any updates here?

same boat, I’d like to use ‘.toHaveStyle()’ in my unit tests.

**Comment by grug** [comment]
We're in the same boat, too. We'd like to test defaults and overrides within our design system but unfortunately can't do it at the moment because of the lack of `toHaveStyle()` functionality.

**Comment by jared-paperspace** [comment]
I had a library for this called [`jest-stitches`](https://github.com/jaredLunde/jest-stitches) but I'm not sure what changed yet that made it break. It doesn't look like it inserts styles to the DOM in the test environment or it does so differently. Working on this tonight if any stitches contributor has a carrot to dangle.

**Reply by grug** [reply]
Hey Jared - if there's anything I can do to help contribute towards this, please let me know. I'm not a Stitches contributor but depending on timezones, I'd be happy to jump on a call and pair program a solution to this with you :)

**Reply by jared-paperspace** [reply]
@grug That would be awesome if we can make it work. I'm in Eastern Standard Time right now. Quick update: they're using [bucketed media queries](https://github.com/modulz/stitches/blob/cb414be79693c783b60923aeda939df6006a0020/packages/core/src/sheet.js#L5-L14) to house styles, but I haven't been able to make them render during tests yet. Posted this to [their discord](https://discord.com/channels/752614004387610674/938498647077834812): 

> Is there something special that has to happen during render with `require('react-test-renderer').create(children).toJSON()` to make the sheets populate? I'm rendering components, the grouped sheets are there, but they aren't getting populated with styles 🤔 `getCssText()` doesn't return anything either

If we can figure that out, the rest is fairly trivial.

**Reply by CaioGrossi** [reply]
Hey guys, I`m facing this same problem. Any updates?

**Comment by ricardoalfaia** [comment]
I was working hard here at the company trying to indicate stitches in our project, but the test issue is making us back away from the difficulty of validating the applied classes as with styled-components using jest-styled-componets. Sad but I'll have to keep styled-components. Everything I see of example today doesn't include unit tests for component stitches. We have this gap for adoption

**Reply by ricardoalfaia** [reply]
Seems like we're all stuck at this point

**Reply by ricardoalfaia** [reply]
it would be nice to have something like .toHaveStyleRule('prop', 'value') or .toHaveStyle(...)

**Reply by mryechkin** [reply]
Yeah, absence of comparable matchers like that for Stitches was basically a non-starter for us. Decided to stick with `styled-components`

**Comment by alvechy** [comment]
I think `toHaveStyle` is not what you really want. I think there are few solutions already that simplify and make visual testing faster, to avoid visual regressions, that work much better than testing arbitrary style rules.
You can always check that the classes are applied correctly through `expect(el).toHaveAttribute('class',  expect.stringContaining('variant-variantName'))` (see [doc](https://github.com/testing-library/jest-dom#tohaveattribute)).
If it's too cumbersome, you can always create a custom matcher.

**Reply by grug** [reply]
It's been a hot minute since I've looked at this, but I seem to recall that the code provided by `jest-dom` doesn't work in conjunction with stitches

**Reply by igortullio** [reply]
This work for me

**Reply by nasimbeem** [reply]
this works for me

**Reply by GilAlvez** [reply]
This worked for me, but I believe it's limited. 
In my case, for example, I have a button component that has many base styles and compoundVariants. The only variation I can test using this method is 'size'. Everything else are unreadable classes, thus untestable.

**Comment by rntvicente** [comment]
Hi all,

Any updates here?

same boat, I’d like to use ‘.toHaveStyle()’ in my unit tests.

**Comment by rahul-budhia** [comment]
Hey everyone,

How do you suggest we test a show / hide toggle functionality?
What I use from jest is [toBeVisible()](https://github.com/testing-library/jest-dom#tobevisible) and `.not.toBeVisible()`, which looks for css prop values like `display` and `visibility`. Since it doesn't work with stitches how do you work around it?

It currently feels like jest's `.not.toBeVisible()` check will not work in any case.

**Comment by hyjoong** [comment]
I'm having the same issues.
I would like to use '.toHaveStyle()' in the unit test, but I have encountered the following issues:

 using `toHaveStyle()`

`Button Component`
```javascript
const Button = ({ isDisabled = false, children, ...props }: ButtonProps) => {
  return (
    <ButtonStyled isDisabled={isDisabled} {...props}>
      {children}
    </ButtonStyled>
  );
};

// button style 
...
        border: "1px solid #00C896",

```
test code
```javascript
it("~~", () => {
    render(<Quiz />);
    const checkButton = screen.getByRole("button", { name: /정답 확인/i });
    expect(checkButton).toHaveStyle(`border: 1px solid #00C896`);
  });
```

<img width="683" alt="image" src="https://user-images.githubusercontent.com/70426440/214030977-a616a4cf-8513-483e-ada3-de2b4c915cf7.png">

I set it to 1px, but I don't know why 2px is coming out.

https://github.com/hyjoong/quiz-app/blob/master/__tests__/quiz.test.jsx

**Comment by jpalvarez** [comment]
In the year 2023, cars drive themselves, and AI rules the world, but we can't do something simple like test a component using "toBeVisible" because stitches don't work (and apparently will never work) with jsdom. Awesome!

**Reply by mryechkin** [reply]
I wouldn't hold my breath:

![Screenshot 2023-11-10 at 1 47 51 PM](https://github.com/stitchesjs/stitches/assets/5657181/d04b30f4-bfed-48a2-a8d5-cf3b998bfc2a)

https://github.com/stitchesjs/stitches/discussions/1149#discussioncomment-6223090
